geomesa-convert/geomesa-convert-common/src/main/scala/org/locationtech/geomesa/convert/SimpleFeatureValidator.scala

```diff
@@ -8,168 +8,256 @@
-        lastErr = "Null geometry attribute"
-        false
-      } else if (date.before(minDate)) {
-        lastErr = s"Date is before Z3 Min Date ($minDate)"
-        false
-      } else if (date.after(maxDate)) {
-        lastErr = s"Date is after Z3 Max Date ($maxDate)"
-        false
-      } else if (!wholeWorldEnvelope.contains(geom.getEnvelopeInternal)) {
-        lastErr = s"Geometry exceeds world bounds ($wholeWorldEnvelope)"
-        false
+  /**
+    * Z3 validator
+    *
+    * @param geom geom index
+    * @param dtg dtg index
+    * @param minDate min z3 date
+    * @param maxDate max z3 date
+    */
+  private class Z3Validator(geom: Int, dtg: Int, minDate: Date, maxDate: Date) extends Validator {
+    private val dateBefore = s"date is before minimum indexable date ($minDate)"
+    private val dateAfter = s"date is after maximum indexable date ($maxDate)"
+
+    override def validate(sf: SimpleFeature): String = {
+      val d = sf.getAttribute(dtg).asInstanceOf[Date]
+      val g = sf.getAttribute(geom).asInstanceOf[Geometry]
+      if (g == null) {
+        if (d == null) {
+          s"${Errors.GeomNull}, ${Errors.DateNull}"
+        } else if (d.before(minDate)) {
+          s"${Errors.GeomNull}, $dateBefore"
+        } else if (d.after(maxDate)) {
+          s"${Errors.GeomNull}, $dateAfter"
+        } else {
+          Errors.GeomNull
+        }
+      } else if (!wholeWorldEnvelope.contains(g.getEnvelopeInternal)) {
+        if (d == null) {
+          s"${Errors.GeomBounds}, ${Errors.DateNull}"
+        } else if (d.before(minDate)) {
+          s"${Errors.GeomBounds}, $dateBefore"
+        } else if (d.after(maxDate)) {
+          s"${Errors.GeomBounds}, $dateAfter"
+        } else {
+          Errors.GeomBounds
+        }
+      } else if (d == null) {
+        Errors.DateNull
+      } else if (d.before(minDate)) {
+        dateBefore
+      } else if (d.after(maxDate)) {
+        dateAfter
       } else {
-        true
+        null
       }
     }
-
-    override def lastError: String = lastErr
   }
 
-  class CompositeValidator(validators: Seq[SimpleFeatureValidator]) extends SimpleFeatureValidator {
-    override def validate(sf: SimpleFeature): Boolean = validators.forall(_.validate(sf))
-    override def init(sft: SimpleFeatureType): Unit = validators.foreach(_.init(sft))
-    override def name: String = validators.map(_.name).mkString(",")
-    override def lastError: String = validators.find(_.lastError != null).map(_.lastError).orNull
+  private object NoValidator extends Validator {
+    override def validate(sf: SimpleFeature): String = null
   }
 
-  object NoneValidator extends SimpleFeatureValidator {
-    override def name: String = "none"
-    override def validate(sf: SimpleFeature): Boolean = true
-    override def init(sft: SimpleFeatureType): Unit = {}
-    override def lastError: String = null
+  private object Errors {
+    val GeomNull   = "geometry is null"
+    val DateNull   = "date is null"
```

Comment:
Idly, I wonder if it is worth naming the field which is null (or out of bounds).

Reply:
the values will be printed out by the converter library, so it should be somewhat apparent... i don't have a strong opinion either way